8675 - user group invites

adiloztaser opened the pull request:
Trac ticket: https://buddypress.trac.wordpress.org/ticket/8675

---
**This Pull Request is for code review only. Please keep all other discussion in the BuddyPress Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the WordPress Core Handbook for more details.**

imath commented on src/bp-templates/bp-nouveau/buddypress/members/single/groups.php:
```diff
@@ -10,7 +10,7 @@
 	<ul class="subnav">
 
-		<?php if ( bp_is_my_profile() ) : ?>
+		<?php if ( bp_core_can_edit_settings() ) : ?>
```
We can simply remove this check actually and simply display the Member's Groups subnav. The Invitations Tab will only show for the users seeing their own profile or for Admins. As you noted into your comment on Track, the Invitations subnav will only show if `bp_core_can_edit_settings()`.

adiloztaser replied:
Good point, thank you.

imath commented on src/bp-templates/bp-legacy/buddypress/members/single/groups.php:
```diff
@@ -11,9 +11,13 @@
+		if ( bp_core_can_edit_settings() ) {
+			bp_get_options_nav();
+		}
+		?>
```
Just use `bp_get_options_nav();` there, as explained in the corresponding comment about the Nouveau Template pack.

adiloztaser replied:
Good point, thank you.

imath reviewed (changes requested):
Thanks a lot for this PR. I agree it's visually fixing the issue. But if an Admin clicks on the button to accept the invite for the user then: 
- we have an Ajax error in Nouveau and the invite is not accepted.
- we have a regular error in Legacy but the invite is accepted.

So changing this behavior needs more changes probably:
- there https://github.com/buddypress/buddypress/blob/master/src/bp-groups/actions/join.php#L29|L33
- and there https://github.com/buddypress/buddypress/blob/master/src/bp-templates/bp-nouveau/includes/groups/ajax.php#L104

But it might not just be a matter of switching from logged in user ID to displayed user ID, it's probably a broader issue.

adiloztaser commented:
Hey @imath, thanks for the review. Actually I've never tried to use `Accept/Reject` buttons, I'll look asap.

adiloztaser commented:
Hi again @imath,

I believe accept/reject buttons should work for admins after my last commits. Can you look again?

I also agree about your concern to inform users but I'm not sure where to generate it.

imath commented on src/bp-groups/bp-groups-template.php:
```diff
@@ -3354,7 +3354,10 @@ function bp_get_group_accept_invite_link( $group = false ) {
 
-		$bp = buddypress();
+		$user_domain = bp_loggedin_user_domain();
+		if ( bp_displayed_user_domain() ) {
```
I think it's best to use `bp_is_user()` instead of `bp_displayed_user_domain()` here

adiloztaser replied:
✅

imath commented on src/bp-groups/bp-groups-template.php:
```diff
@@ -3392,7 +3395,10 @@ function bp_get_group_reject_invite_link( $group = false ) {
 
-		$bp = buddypress();
+		$user_domain = bp_loggedin_user_domain();
+		if ( bp_displayed_user_domain() ) {
```
I think it's best to use `bp_is_user()` instead of `bp_displayed_user_domain()` here

adiloztaser replied:
✅

imath reviewed (changes requested):
Hi @adiloztaser 

Thanks a lot for your changes 👍. I've just tested, it works as expected. About emailing the user, i'd use the `'groups_reject_invite'`  & `'groups_accept_invite'` hook to check `if ( $user_id !== bp_loggedin_user_id() && bp_current_user_can( 'bp_moderate' )` to decide whether to send a notification email.

adiloztaser commented:
Hi again @imath,

I've added an email notification based on your suggestion.
I have two questions:
1-) Should we create a new email or just use existing ones?
2-) Should we add unsubscribe link to email or we always inform users after that action?

imath commented:
Thanks a lot @adiloztaser for these improvements.

I think we should have two specific emails eg: "An administrator accepted an invitation to join {{group.name}} on your behalf. If you disagree with this, you can leave the group at anytime using this link: {{{leave-group.url}}}

In this case we don't need the unsubscribe link.

The new code you added looks great.

adiloztaser commented:
Thank you for feedback @imath, I'll keep working on it.

adiloztaser commented on src/bp-groups/bp-groups-notifications.php:
```diff
@@ -1222,3 +1222,42 @@ function groups_screen_notification_settings() {
+		return;
+	}
+
+	$group = groups_get_group( $group_id );
```
Thank you. ✅

adiloztaser commented on src/bp-groups/bp-groups-notifications.php:
```diff
@@ -1222,3 +1222,42 @@ function groups_screen_notification_settings() {
+ * @param int $group_id ID of the group.
+ */
+function groups_email_notification_membership_request_completed_by_admin( $user_id, $group_id ) {
+	if ( $user_id === bp_loggedin_user_id() ) {
```
Thank you. ✅

adiloztaser commented:
Hi there.

I've added new emails, as you guess to do that I needed to increase the `db_version`. I don't know is there any rule for this, if I did anything please let me know.

I couldn't add leave url to accepted invate email because we have nonce check and I didn't create a valid url. I've added member groups page, I think it's good enough for navigating user.

imath reviewed (changes requested):
It's looking great thanks a lot for you work on it @adiloztaser 

I just think we should do the small changes I suggested in my review 😉

adiloztaser commented on src/bp-core/bp-core-functions.php:
```diff
@@ -3971,6 +3971,22 @@ function bp_email_get_schema() {
 		),
+		'groups-membership-request-accepted-by-admin' => array(
+			/* translators: do not remove {} brackets or translate its contents. */
+				'post_title'   => __( '[{{{site.name}}}] Membership request for group "{{group.name}}" accepted', 'buddypress' ),
```
Thank you, it looks like I need to fix my editor config :facepalm:

adiloztaser commented on src/bp-core/bp-core-functions.php:
```diff
@@ -3971,6 +3971,22 @@ function bp_email_get_schema() {
+		),
+		'groups-membership-request-rejected-by-admin' => array(
+			/* translators: do not remove {} brackets or translate its contents. */
+				'post_title'   => __( '[{{{site.name}}}] Membership request for group "{{group.name}}" rejected', 'buddypress' ),
```
:facepalm:

adiloztaser commented on src/bp-core/bp-core-functions.php:
```diff
@@ -3971,6 +3971,22 @@ function bp_email_get_schema() {
+			/* translators: do not remove {} brackets or translate its contents. */
+				'post_title'   => __( '[{{{site.name}}}] Membership request for group "{{group.name}}" rejected', 'buddypress' ),
+			/* translators: do not remove {} brackets or translate its contents. */
+				'post_content' => __( "An administrator rejected an invitation to join &quot;<a href=\"{{{group.url}}}\">{{group.name}}</a>&quot; on your behalf.\n\nIf you disagree with this, please contact the site administrator.", 'buddypress' ),
```
:facepalm:

adiloztaser commented on src/bp-core/bp-core-functions.php:
```diff
@@ -4152,6 +4168,18 @@ function bp_email_get_type_schema( $field = 'description' ) {
+	);
+
+	$groups_membership_request_rejected_by_admin = array(
+			'description'	   => __( 'Recipient had requested to join a group, which was rejected by admin.', 'buddypress' ),
+			'named_salutation' => true,
+			'unsubscribe'	   => false,
```
:facepalm:

adiloztaser commented on src/bp-core/bp-core-update.php:
```diff
@@ -284,6 +284,11 @@ function bp_version_updater() {
 		}
 
+		// Version 11.0.0.
+		if ( $raw_db_version < 14030 ){
```
:+1: Thanks for the information.

adiloztaser commented:
Thank you for review and suggestions @imath. I hope its ready to go.

imath reviewed (approved):
Looks good! Thanks a lot @adiloztaser I'll commit it to our SVN Trunk asap.
